### PR TITLE
Fix warnings with Python 3.10

### LINKED ISSUE #11821: Test faillures with python 3.10.0a6
Fedora is starting to build python packages with Python 3.10.0a6 in its development version in order to detect problems early.

There are a few errors (very few) in testing astropy with pytest during the building process. 
Some of them seem related to the handling of errors (a number of them are expected to be captured but a different number of them appear).

```
=================================== FAILURES ===================================
___________________ TestFileFunctions.test_mmap_unwriteable ____________________

self = <astropy.io.fits.tests.test_core.TestFileFunctions object at 0x7fcdda8732e0>

    def test_mmap_unwriteable(self):
        """Regression test for https://github.com/astropy/astropy/issues/968
    
        Temporarily patches mmap.mmap to exhibit platform-specific bad
        behavior.
        """
    
        class MockMmap(mmap.mmap):
            def flush(self):
                raise OSError('flush is broken on this platform')
    
        old_mmap = mmap.mmap
        mmap.mmap = MockMmap
    
        # Force the mmap test to be rerun
        _File.__dict__['_mmap_available']._cache.clear()
    
        try:
            self.copy_file('test0.fits')
            with pytest.warns(AstropyUserWarning, match=r'mmap\.flush is unavailable') as w:
                with fits.open(self.temp('test0.fits'), mode='update',
                               memmap=True) as h:
                    h[1].data[0, 0] = 999
    
>           assert len(w) == 1
E           assert 4 == 1
E            +  where 4 = len(WarningsChecker(record=True))

astropy/io/fits/tests/test_core.py:1007: AssertionError
___________________ TestHDUListFunctions.test_flush_readonly ___________________

self = <astropy.io.fits.tests.test_hdulist.TestHDUListFunctions object at 0x7fcdf8169300>

    def test_flush_readonly(self):
        """Test flushing changes to a file opened in a read only mode."""
    
        oldmtime = os.stat(self.data('test0.fits')).st_mtime
        with fits.open(self.data('test0.fits')) as hdul:
            hdul[0].header['FOO'] = 'BAR'
            with pytest.warns(AstropyUserWarning, match='mode is not supported') as w:
                hdul.flush()
>           assert len(w) == 1
E           assert 4 == 1
E            +  where 4 = len(WarningsChecker(record=True))

astropy/io/fits/tests/test_hdulist.py:462: AssertionError
___________________ TestHDUListFunctions.test_output_verify ____________________

self = <astropy.io.fits.tests.test_hdulist.TestHDUListFunctions object at 0x7fcdfdbe36d0>

    def test_output_verify(self):
        hdul = fits.HDUList([fits.PrimaryHDU()])
        hdul[0].header['FOOBAR'] = 42
        hdul.writeto(self.temp('test.fits'))
    
        with open(self.temp('test.fits'), 'rb') as f:
            data = f.read()
        # create invalid card
        data = data.replace(b'FOOBAR  =', b'FOOBAR = ')
        with open(self.temp('test2.fits'), 'wb') as f:
            f.write(data)
    
        with pytest.raises(VerifyError):
            with fits.open(self.temp('test2.fits'), mode='update') as hdul:
                hdul[0].header['MORE'] = 'here'
    
        with pytest.warns(VerifyWarning) as ww:
            with fits.open(self.temp('test2.fits'), mode='update',
                           output_verify='fix+warn') as hdul:
                hdul[0].header['MORE'] = 'here'
>       assert len(ww) == 6
E       assert 9 == 6
E        +  where 9 = len(WarningsChecker(record=True))

astropy/io/fits/tests/test_hdulist.py:1083: AssertionError
_________________________ test_read_write_format[json] _________________________

fmt = 'json'

    @pytest.mark.parametrize('fmt', WRITE_FMTS)
    def test_read_write_format(fmt):
        """
        Test round-trip through pandas write/read for supported formats.
    
        :param fmt: format name, e.g. csv, html, json
        :return:
        """
        # Skip the reading tests
        if fmt == 'html' and not HAS_HTML_DEPS:
            pytest.skip('Missing lxml or bs4 + html5lib for HTML read/write test')
    
        pandas_fmt = 'pandas.' + fmt
        # Explicitly provide dtype to avoid casting 'a' to int32.
        # See https://github.com/astropy/astropy/issues/8682
        t = Table([[1, 2, 3], [1.0, 2.5, 5.0], ['a', 'b', 'c']],
                  dtype=(np.int64, np.float64, str))
        buf = StringIO()
>       t.write(buf, format=pandas_fmt)

astropy/io/misc/tests/test_pandas.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/table/connect.py:127: in __call__
    registry.write(instance, *args, **kwargs)
astropy/io/registry.py:563: in write
    writer(data, *args, **kwargs)
astropy/io/misc/pandas/connect.py:113: in _pandas_write
    return write_method(filespec, **write_kwargs)
/usr/lib64/python3.10/site-packages/pandas/core/generic.py:2465: in to_json
    return json.to_json(
/usr/lib64/python3.10/site-packages/pandas/io/json/_json.py:95: in to_json
    ).write()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pandas.io.json._json.FrameWriter object at 0x7fcddac6b670>

    def write(self):
        iso_dates = self.date_format == "iso"
>       return dumps(
            self.obj_to_write,
            orient=self.orient,
            double_precision=self.double_precision,
            ensure_ascii=self.ensure_ascii,
            date_unit=self.date_unit,
            iso_dates=iso_dates,
            default_handler=self.default_handler,
            indent=self.indent,
        )
E       ValueError: "dtype('O')" or "<class 'pandas.core.indexes.base.Index'>" are not JSON serializable yet

/usr/lib64/python3.10/site-packages/pandas/io/json/_json.py:147: ValueError
____________________________ test_warnings_logging _____________________________

    def test_warnings_logging():
    
        # Without warnings logging
        with pytest.warns(AstropyUserWarning, match="This is a warning") as warn_list:
            with log.log_to_list() as log_list:
                warnings.warn("This is a warning", AstropyUserWarning)
        assert len(log_list) == 0
        assert len(warn_list) == 1
    
        # With warnings logging
        with pytest.warns(None) as warn_list:
            log.enable_warnings_logging()
            with log.log_to_list() as log_list:
                warnings.warn("This is a warning", AstropyUserWarning)
            log.disable_warnings_logging()
        assert len(log_list) == 1
>       assert len(warn_list) == 0
E       assert 2 == 0
E        +  where 2 = len(WarningsChecker(record=True))

astropy/tests/test_logger.py:89: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: This is a warning [astropy.tests.test_logger]
------------------------------ Captured log call -------------------------------
WARNING  astropy:logger.py:204 This is a warning
___________________ test_warnings_logging_with_custom_class ____________________

    def test_warnings_logging_with_custom_class():
        class CustomAstropyWarningClass(AstropyWarning):
            pass
    
        # With warnings logging
        with pytest.warns(None) as warn_list:
            log.enable_warnings_logging()
            with log.log_to_list() as log_list:
                warnings.warn("This is a warning", CustomAstropyWarningClass)
            log.disable_warnings_logging()
        assert len(log_list) == 1
>       assert len(warn_list) == 0
E       assert 2 == 0
E        +  where 2 = len(WarningsChecker(record=True))

astropy/tests/test_logger.py:127: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: CustomAstropyWarningClass: This is a warning [astropy.tests.test_logger]
------------------------------ Captured log call -------------------------------
WARNING  astropy:logger.py:204 CustomAstropyWarningClass: This is a warning
_________________ test_warning_logging_with_io_votable_warning _________________

    def test_warning_logging_with_io_votable_warning():
        from astropy.io.votable.exceptions import W02, vo_warn
    
        with pytest.warns(None) as warn_list:
            log.enable_warnings_logging()
            with log.log_to_list() as log_list:
                vo_warn(W02, ('a', 'b'))
            log.disable_warnings_logging()
        assert len(log_list) == 1
>       assert len(warn_list) == 0
E       assert 2 == 0
E        +  where 2 = len(WarningsChecker(record=True))

astropy/tests/test_logger.py:142: AssertionError
```

### PR DESCRIPTION
**Pull request by saimn**
Closes  #11821, tests run locally, not sure when we will be able to add a job on Actions.

Fix the deprecation warnings and ignore new `RuntimeWarning` raised in convolution, lombscargle, modeling and units. We should probably address those by either fixing the code or catching them in tests but there are many of them (except for units, only 1 IIRC). So probably to address those when we have a 3.10 job.

What should be the milestone, probably 4.3 ?

### Description

This pull request is to address ...

Fixes #<Issue Number>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/convolution/convolve.py
Comment (pllim) [inline astropy/convolution/convolve.py:29]: This fix does not look like it is going to work when the deprecated feature is removed. Is there no replacement for this?

File: astropy/utils/introspection.py
Comment (pllim) [inline astropy/utils/introspection.py:130]: Oh no... How do we replace this?

File: setup.cfg
Comment (pllim) [inline setup.cfg:150]: I am not comfortable with blanket ignore like this. They should be ignored on a case-by-case basis.

File: astropy/convolution/convolve.py
Comment (saimn) [reply astropy/convolution/convolve.py:29]: Numpy has some plan : https://github.com/numpy/numpy/issues/18588
But here we are using a smaller helper function that relies on distutils to get the shared library extension on different platforms. We could replace this with some custom code I guess.

File: setup.cfg
Comment (saimn) [reply setup.cfg:150]: Yes, but it's more work, and at least for lombscargle I have no idea how to fix it.

File: astropy/convolution/convolve.py
Comment (pllim) [reply astropy/convolution/convolve.py:29]: So, instead of merging this, which would completely mask the real problem, should we wait a bit and see? Not like we can test against Python 3.10 in CI right now anyway...

**Comment by saimn** [comment]
Just as a remainder, Python 3.10's release schedule:
```
    3.10.0 beta 4: Saturday, 2021-07-10

Expected:

    3.10.0 candidate 1: Monday, 2021-08-02
    3.10.0 candidate 2: Monday, 2021-09-06 (if necessary)
    3.10.0 final: Monday, 2021-10-04
```

File: astropy/convolution/convolve.py
Comment (saimn) [reply astropy/convolution/convolve.py:29]: I don't think they will fix this in a few weeks, and we/they have time until 3.12
Actions seems to support 3.10b4 now so I'm trying to add a job.

File: setup.cfg
Comment (pllim) [reply setup.cfg]: I still have a bad feeling about this. This will ignore such warnings for the entire repo, not just the affected subpackages.

File: astropy/convolution/convolve.py
Comment (pllim) [reply astropy/convolution/convolve.py:29]: Okay, as long as we open follow-up issue about this, I guess I can live with this for now.

**Comment by saimn** [comment]
For the record, `convolution` has 573 failed tests, I think all because of the same warning:
```
>               normalized_kernel = kernel / kernel_scale
E               RuntimeWarning: invalid value encountered in true_divide

astropy/convolution/convolve.py:691: RuntimeWarning
```
where `kernel_scale` is `(1+0j)`.

`modeling` has only one failing test, also because of a division by `0j`.

`lombscargle` fails on something like `numerator / denominator` where `numerator` is complex and `denominator` is real, so again division by 0j but not the same warning ...

**Comment by saimn** [comment]
And the 3.10 job passes :tada:

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/11962#issuecomment-883552780 -- Can we localize the warning ignore to the places you have identified instead?

File: setup.cfg
Comment (pllim) [inline setup.cfg:151]: Okay, I guess this is a good compromise. 😸

File: .github/workflows/ci_workflows.yml
Comment (pllim) [inline .github/workflows/ci_workflows.yml:61]: Originally when I attempted this, I put this in devdeps job.

I also wonder if "minimal dependencies" is enough to catch everything...

Comment (pllim) [reply .github/workflows/ci_workflows.yml:61]: Also would need a follow up issue to test with proper Python 3.10 when it is released for real.

File: astropy/units/core.py
Comment (mhvk) [inline astropy/units/core.py:476]: Hmm, I'd be interested to know what the actual warning was. I cannot find this in #11821. Is there perhaps a log I can look at? (I've yet to install a beta version of python; if that's easy, I'll do that!)

File: astropy/utils/introspection.py
Comment (mhvk) [reply astropy/utils/introspection.py:130]: Yes, it seems time to get back to that one!!

File: astropy/convolution/convolve.py
Comment (mhvk) [reply astropy/convolution/convolve.py:29]: Yes, hopefully this can be an "upstream fix required"!

**Review by mhvk — COMMENTED** [review]
Not bad, actually! Overall, looks good. I'm surprised by the warning on the one quantity test, though...

File: .github/workflows/ci_workflows.yml
Comment (saimn) [reply .github/workflows/ci_workflows.yml:61]: Yes we'll need to update for rc1 and the final release.
I didn't try with all other dependencies because I'm afraid this will fail somewhere. But it's fine for now, I mean the goal here is to test that we are compatible with 3.10, not to test that other dependencies also are.

File: astropy/units/core.py
Comment (saimn) [reply astropy/units/core.py:476]: Same problem as all others failures, dividing a complex by a scalar now raises a warning because of the division by `0j`:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
465         list of equivalent pairs, e.g., as returned by
466         `~astropy.units.equivalencies.dimensionless_angles`.
467 
468     Examples
469     --------
470     Exponentiation normally requires dimensionless quantities.  To avoid
471     problems with complex phases::
472 
473         >>> from astropy import units as u
474         >>> with u.set_enabled_equivalencies(u.dimensionless_angles()):
UNEXPECTED EXCEPTION: RuntimeWarning('divide by zero encountered in cdouble_scalars')
Traceback (most recent call last):
  File "/home/sconseil/.pyenv/versions/3.10.0b4/lib/python3.10/doctest.py", line 1348, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest astropy.units.core.set_enabled_equivalencies[1]>", line 3, in <module>
  File "/home/sconseil/dev/astropy/astropy/astropy/units/quantity.py", line 504, in __array_ufunc__
    arrays.append(converter(input_) if converter else input_)
  File "/home/sconseil/dev/astropy/astropy/astropy/units/core.py", line 988, in convert
    return func(_condition_arg(v) / scale1) * scale2
RuntimeWarning: divide by zero encountered in cdouble_scalars
/home/sconseil/dev/astropy/astropy/astropy/units/core.py:474: UnexpectedException
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/sconseil/dev/astropy/astropy/astropy/units/core.py(988)convert()
-> return func(_condition_arg(v) / scale1) * scale2
(Pdb) _condition_arg(v)
0.5j
(Pdb) scale1
0.15915494309189535
(Pdb) _condition_arg(v) / scale1
*** RuntimeWarning: divide by zero encountered in cdouble_scalars
```

**Review by pllim — APPROVED** [review]
LGTM though we need to open several follow-up issues.

I won't merge yet in case subpackage maintainers want to review all the warnings in `modeling` (@nden , @perrygreenfield), `timeseries` (@astrofrog, @bsipocz), `units` (@adrn, @mhvk), and `convolution` (@larrybradley, @adonath, @keflavich).

File: .github/workflows/ci_workflows.yml
Comment (bsipocz) [reply .github/workflows/ci_workflows.yml:61]: Long term I think it's worth bundling this with dev dependencies, especially if they provide a nightly build with dev/rc python.

**Review by bsipocz — APPROVED** [review]

**Comment by saimn** [comment]
To sum up, all warnings are caused by some kind of division of a complex by a scalar:
```
>>> np.complex128(1j) / 2
<stdin>:1: RuntimeWarning: divide by zero encountered in cdouble_scalars
0.5j

>>> np.array([1j]) / np.array([2])
<stdin>:1: RuntimeWarning: divide by zero encountered in true_divide
array([0.+0.5j])

>>> np.array([1 + 0j]) / 1+0j
<stdin>:1: RuntimeWarning: invalid value encountered in true_divide
array([1.+0.j])
```

File: astropy/units/core.py
Comment (mhvk) [reply astropy/units/core.py:476]: Is this python 3.10 specific? This is completely unphysical, so I'd want to raise a bug report...

**Comment by mhvk** [comment]
Weird, so it is a numpy problem, but one that only comes up in python 3.10? I just installed python 3.10 via debian experimental, and indeed cannot reproduce that error with just `0.5j/2.`, so it is not directly a python problem. Unfortunately, I cannot seem to use this installation with `venv` or `virtualenv`, so it is not much use for further testing... But I did check that I do not get your warnings on numpy-dev/python 3.9. 

One question: perhaps we should try with numpy-dev here as well? They seem to have made quite a few fixes for python 3.10 (though nothing that looks really relevant...). If this also gives problems with numpy-dev, they really should know about this...

**Comment by saimn** [comment]
Yes, it only happens on Python 3.10 and with Numpy complex values, not with Python ones. Maybe we can try with Numpy's nightly.

**Comment by pllim** [comment]
`ERROR: Could not find a version that satisfies the requirement numpy (from versions: none)`

Looks like Numpy dev not compatible with Python 3.10? 🤷

**Comment by saimn** [comment]
No, probably they do not build the nightly wheels for 3.10 yet.

**Comment by pllim** [comment]
Hmm passed on Python 3.10 and numpy-dev, it seems. But that is with the ignore still in place, right?

**Comment by saimn** [comment]
I commented the filters in the previous commit, so yeah it seems to be OK with numpy dev !
Maybe I revert the two last commits, and we remove the filters when the next version of numpy is out ?

**Comment by pllim** [comment]
Or... we just stick Python 3.10 into devdeps job and be done with it?

**Comment by saimn** [comment]
Installing Numpy from the git repo makes the job longer, compiling Numpy. It's an option but we can also wait for Numpy to publish nightly wheels for 3.10.

**Comment by bsipocz** [comment]
> Installing Numpy from the git repo makes the job longer, compiling Numpy. It's an option but we can also wait for Numpy to publish nightly wheels for 3.10.

right now I would add this to the cron tests, as well as to the extras and would switch to the default devjob once the nightlies are available.

**Comment by pllim** [comment]
I like @bsipocz 's idea to put this in cron! I think it is easier to follow up on than to ignore warnings and then un-ignore.

**Comment by saimn** [comment]
Ok, seems a good compromise. I'm pushing the changes but still with the normal job because I had removed matplotlib-dev, so better to test if it compile here before relying on cron.

**Comment by saimn** [comment]
Hmm, asdf-dev tests are failing ...

**Comment by mhvk** [comment]
Thanks all for looking into it!  Using 3.10 with numpy-dev indeed sounds like the right solution - it is good to see that this will make the number of problems (even) smaller!

**Comment by pllim** [comment]
`asdf.resource_mappings plugin from package asdf==2.9.0.dev6+g887c5be failed to load` -- @eslavich , any idea?

**Comment by saimn** [comment]
3.10rc1 is out so I updated the version used in Actions, but asdf-dev is still failing. How shall we proceed ?

**Comment by pllim** [comment]
Re: asdf-dev -- @eslavich or @jdavies-st , any ideas?

**Comment by jdavies-st** [comment]
> `asdf.resource_mappings plugin from package asdf==2.9.0.dev6+g887c5be failed to load` -- @eslavich , any idea?

Can you point us to a log or traceback somewhere?

**Comment by saimn** [comment]
>  From your tox file it looks like everything is installed into site-packages.

Yes, I think so.

**Comment by saimn** [comment]
Numpy now provides wheels for 3.10 so I updated the branch and it seems the asdf-dev error is gone. So I think this is ready to go ?

File: .github/workflows/ci_cron_weekly.yml
Comment (astrofrog) [inline .github/workflows/ci_cron_weekly.yml:34]: I think there is a syntax that allows you to say 'or later 3.10 releases'? If you can't find it I can look when I am back at a computer today.

Comment (astrofrog) [reply .github/workflows/ci_cron_weekly.yml:34]: ´´´
python-version: '3.10.0-alpha - 3.10.0 # SemVer's version range syntax
´´´

Comment (saimn) [reply .github/workflows/ci_cron_weekly.yml]: Done, thanks!

**Comment by pllim** [comment]
@saimn , there is conflict now. FYI

**Comment by saimn** [comment]
All green now :)

**Review by pllim — APPROVED** [review]
All the cron stuff are broken for different and unrelated reasons, so no point trying to run them here. Let's get this in. Thanks!

**Comment by pllim** [comment]
ops... Do we need to backport?

**Comment by saimn** [comment]
Having a 4.3.x compatible with Python 3.10 would be nice (it will out in ~1 month). I will look at the manual backport later.

**Comment by vstinner** [comment]
Thanks for the fix @saimn!

**Pull request merged**
